src/core/electrostatics_magnetostatics/p3m.cpp

```diff
@@ -707,12 +705,11 @@ static double p3m_mcr_time(const int mesh[3], int cao, double r_cut_iL,
  *           -@ref P3M_TUNE_FAIL, -@ref P3M_TUNE_ACCURACY_TOO_LARGE,
  *           -@ref P3M_TUNE_CAO_TOO_LARGE, or -@ref P3M_TUNE_ELCTEST
  */
-static double p3m_mc_time(char **log, const int mesh[3], int cao,
-                          double r_cut_iL_min, double r_cut_iL_max,
-                          double *_r_cut_iL, double *_alpha_L,
-                          double *_accuracy) {
+template <bool verbose>
+static double p3m_mc_time(const int mesh[3], int cao, double r_cut_iL_min,
```

Comment:
Here and elsewhere: I do not immediately see the advantage of using a template parameter instead of a regular parameter. Could you please let me know why this was necessary?

With a template function, there is no guarantee the `true` and `false` instantiations will do the same calculation. It is, for example, perfectly legal for someone to inject this snippet of code at line 798:
```c++
template <>
double p3m_mc_time<false>(const int mesh[3], int cao, double r_cut_iL_min,
                          double r_cut_iL_max, double *_r_cut_iL,
                          double *_alpha_L, double *_accuracy) {
  return 0.;
}
```
This will cause the tuning to select random values, which means we should duplicate the mmm1d/p3m/dp3m python tests to make sure `verbose=False` gives the same results as `verbose=True`.

Reply:
The template was just personal preference. I dont care, your choice is probably to change it to a regular paramter, i can quickly do that if you want me to.

Reply:
I think it would be more maintainable with a regular parameter. This would also allow you to merge `p3m_adaptive_tune_core()` and `p3m_adaptive_tune()` back into one function.

Reply:
im working on it, this is an artifact of cython not supporting non-type templates